### DISCUSSION TITLE
Custom Post Types

### DISCUSSION THREAD

**Discussion by HitArrowLegend** [GinoPane/oc-blogtaxonomy-plugin#85] [Ideas]
Hey Gino, 

I was wondering if custom types would be something you're interested in working on to add to your addon. I have ideas on the subject that I would really love to have on taxonomy; Possibly as custom work ?

Thinking about Review Posts mainly, but this could be extended so much more.

**Comment by GinoPane** [comment]
Hi!

Do you have more examples maybe? Isn't it possible to achieve the same using categories? And how types are going to be different to categories/tags? Should it be possible to add new custom types?

I thought about something similar in context of implementation translation posts, but then started working on something else and forgot about the idea.

**Reply by HitArrowLegend** [reply]
Hello !

Categories, tags and series are indeed very useful to have a clean and organized blog, but what I mean goes beyond that.

As an example, a video game review could have a category,a tag and belong to a series of posts, but it will also have elements outside the scope of just plain text. The developer, the release date, the editor, the rating etc... are all datas that could appear in every video game review you would write on your blog. 
A custom type dedicated to TV Shows could have the list of actors on every article associated and so on.
Needless to say that this could help further expand the uniqueness of every blogs.

Check this article : https://www.ign.com/articles/valheim-is-taking-over-steam-and-twitch-despite-being-in-early-access 
At the bottom of the article, you have these elements that are on every other articles, giving us the name of the game, the publisher, the editor, the release date, the platforms its on. 
On this one, you have those fingers that count as a rating, but it also has a category (it's a review) and a tag, (lamb of god).
https://www.metalsucks.net/2020/06/17/album-review-do-lamb-of-god-go-all-black-album-on-lamb-of-god/
This is what I'd like to be able to edit with custom types on Taxonomy.

About the current components, I would just imagine having new filters, one to exclude certain custom types, or to allow only certain ones (to have a widget full of reviews for example).

The way it could work is:

- User create a custom post type : Music Album Review
- User add custom fields to this new type such as a Rating, the Release Date, the Genre, the Record Label etc...
- User create a new page dedicated to this custom post type (and decorate it the way he wants)
- User create a post, adds it the custom type, fills the blanks, adds a tag, a categories, and voila.

I could make some photoshop edits to maybe give a better understanding of what I mean.

**Reply by GinoPane** [reply]
Ok, I see. This is what I thought about too. 
So basically post type should provide some additional properties for posts?

Admin creates a post, chooses a type for it and it makes some additional fields available for setup?

**Reply by HitArrowLegend** [reply]
Yes, exactly how you said it

**Reply by GinoPane** [reply]
@HitArrowLegend from the first glance I see no easy way to implement this in October CMS. I mean, it may be possible to do this via framework features, but definitely requires some more investigation.

So if you are good at October and has maybe proposals regarding implementation (or even pull requests) it would be great.

**Comment by GinoPane** [comment]
@HitArrowLegend I created a dev ticket related to this #87. Please take a look and let me know if it is what you expected. 

Also, I'm planning to make some initial type available as an example. This could be a "review" type as you mentioned before, but I need a proper set of initial fields for that. Of course they will be customizable, but we need something as a starter.

**Reply by HitArrowLegend** [reply]
I just checked your message. I'll go with some points I'm not sure to fully understand.

"Post types would be an additional model with its own backend lists, views, etc."

Does that mean that we would have posts, and post types as two different entities ? How would that work with your current components and the ones from Rainlab Blog ? If it is easy to make sure both Posts and those new models can coexist seemlessly, then yeah, this is a great solution. 

Everything else I think is great.

Edit: Going through your message about initial fields, I was thinking if it was part of your plans to have those fields acting the way tags do. For instance, Adding an Actor field to a movie review would allow us to add Actors, but then to just type first letters and see those that match what we type in that field.

**Reply by HitArrowLegend** [reply]
About some initial fields for the review post types, I compiled some "basic" fields that you can chose from as a starter. 

**Video Game Reviews**

- Release Date
- Publisher
- Editor
- Rating
- Pros
- Cons

**Music Album Reviews**

- Rating
- Release Date
- Genre
- Label
- Album Artwork

**Movie Reviews**

- Release Date
- Length
- Genre
- Rating
- Main Actors

**Reply by GinoPane** [reply]
> I just checked your message. I'll go with some points I'm not sure to fully understand.
> 
> "Post types would be an additional model with its own backend lists, views, etc."
> 
> Does that mean that we would have posts, and post types as two different entities ? How would that work with your current components and the ones from Rainlab Blog ? If it is easy to make sure both Posts and those new models can coexist seemlessly, then yeah, this is a great solution.
> 
> Everything else I think is great.
> 
> Edit: Going through your message about initial fields, I was thinking if it was part of your plans to have those fields acting the way tags do. For instance, Adding an Actor field to a movie review would allow us to add Actors, but then to just type first letters and see those that match what we type in that field.

@HitArrowLegend  Answering on your edited question: unfortunately, no. It would be too tricky to create models on the fly 😃 
Right now it works like this (we have a dropdown with types and attributes to fill depending on the type; types on their own are edited as separated models):

![image](https://user-images.githubusercontent.com/3897579/110684009-af1bd080-81ed-11eb-9ffd-2c740f7f78cd.png)

![image](https://user-images.githubusercontent.com/3897579/110684263-fc983d80-81ed-11eb-9900-fdea9ce0d717.png)

**Reply by HitArrowLegend** [reply]
looks awesome (and that kitty is very cute)

**Comment by GinoPane** [comment]
@HitArrowLegend the plugin version 3.0.0 has been released. I didn't feel comfortable to postpone it even more, because there were no much active development on it. 
Please report any bugs/issues/suggestions.

**Reply by HitArrowLegend** [reply]
I'll update it but all i tested so far is working great
